dotCMS/src/main/java/com/dotmarketing/business/web/HostWebAPIImpl.java

```diff
@@ -52,49 +57,97 @@ public Host getHost(HttpServletRequest request)  {
-            } else {
-
-                final String serverName = request.getServerName();
-                host = resolveHostName(serverName, systemUser, respectFrontendRoles);
-            }
+        Optional<Host> optionalHost = this.getCurrentHostFromRequest(request, user, respectAnonPerms);
+
+        if (!optionalHost.isPresent() && user.isBackendUser()){
+            optionalHost = this.getCurrentHostFromSession(request, user, respectAnonPerms);
         }
 
+        final Host host = optionalHost.isPresent() ? optionalHost.get() : resolveHostName(request.getServerName(),
+                user, respectAnonPerms);
+
+        checkHostPermission(user, respectAnonPerms, host);
+        storeCurrentHost(request, user, host);
+
+        return host;
+    }
+
+    private void storeCurrentHost(final HttpServletRequest request, final User user, final Host host) {
+        final HttpSession session = request.getSession(false);
+
         request.setAttribute(WebKeys.CURRENT_HOST, host);
-        if (session != null && mode.isAdmin) {
 
+        if (session != null && user.isBackendUser()) {
             session.setAttribute(WebKeys.CURRENT_HOST, host);
         }
+    }
 
-        return host;
+    private void checkHostPermission(User user, boolean respectAnonPerms, Host host) throws DotDataException, DotSecurityException {
+        if(!APILocator.getPermissionAPI().doesUserHavePermission(host, PermissionAPI.PERMISSION_READ, user, respectAnonPerms)){
+            String u = (user != null) ? user.getUserId() : null;
+
+            String message = "User " + u + " does not have permission to host:" + host.getHostname();
+            Logger.error(HostAPIImpl.class, message);
+            throw new DotSecurityException(message);
+        }
     }
 
-	@Override
+    private Optional<Host> getCurrentHostFromSession(final HttpServletRequest request, User user, boolean respectAnonPerms)
+            throws DotSecurityException, DotDataException {
+
+        final HttpSession session = request.getSession(false);
+
+        if (session == null){
+            return Optional.empty();
+        }
+
+        Host host = null;
+        if (session.getAttribute(WebKeys.CURRENT_HOST) != null) {
+            final Host hostFromSession = (Host) session.getAttribute(WebKeys.CURRENT_HOST);
+            final Object hostId = session.getAttribute(WebKeys.CMS_SELECTED_HOST_ID);
+            host = hostFromSession.getIdentifier().equals(hostId) ? hostFromSession : null;
+        }
+
+        if (host == null && session.getAttribute(WebKeys.CMS_SELECTED_HOST_ID) != null) {
+            final String hostId = (String) session.getAttribute(WebKeys.CMS_SELECTED_HOST_ID);
+            host = find(hostId, user, respectAnonPerms);
+        }
+
+        return Optional.ofNullable(host);
+    }
+
+    private Optional<Host> getCurrentHostFromRequest(final HttpServletRequest request, final User user, boolean respectAnonPerms)

```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'respectAnonPerms' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4812018)